clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1034,28 +1024,35 @@ private Record nextFetchedRecord() {
                     records = currentBatch.streamingIterator();
                 }
 
-                Record record = records.next();
-                maybeEnsureValid(record);
-
-                // skip any records out of range
-                if (record.offset() >= nextFetchOffset) {
-                    nextFetchOffset = record.offset() + 1;
-
-                    // control records are not returned to the user
-                    if (!currentBatch.isControlBatch())
-                         return record;
+                // If an exception was thrown from the last record. We should throw the same exception here again.
+                // We do not cache the exception thrown last time because the stack trace could be different
```

Comment:
I think caching the exception may actually be OK. For both `SerializationException` and `InvalidRecordException`, we've overridden `fillInStackTrace` anyway.